README.md
# [Equator](http://equator.com)

by [David Amirkhanov](http://github.com/websofter) (<mail.websofter@gmail.com>) [<img alt="slackin.equator.com" src="http://slackin.equator.com/badge.svg" align="top">](http://slackin.equator.com)

Equator is a web formula editor designed to make typing math easy and beautiful.

[<img alt="homepage demo" src="https://cloud.githubusercontent.com/assets/225809/15163580/1bc048c4-16be-11e6-98a6-de467d00cff1.gif" width="260">](http://equator.com)

The Equator project is supported by its [partners](http://equator.com/partners.html). We hold ourselves to a compassionate [Code of Conduct](http://docs.equator.com/en/latest/Code_of_Conduct/).

Equator is resuming active development and we're committed to getting things running smoothly. Find a dusty corner? [Let us know in Slack.](http://slackin.equator.com) (Prefer IRC? We're `#equator` on Freenode.)

## Getting Started

Equator has a simple interface. This brief example creates a Equator element and renders, then reads a given input:
```javascript
var htmlElement = document.getElementById('some_id');
var config = {
  handlers: { edit: function(){ ... } },
  restrictMismatchedBrackets: true
};
var mathField = MQ.MathField(htmlElement, config);

mathField.latex('2^{\\frac{3}{2}}'); // Renders the given LaTeX in the Equator field
mathField.latex(); // => '2^{\\frac{3}{2}}'
```

Check out our [Getting Started Guide](http://docs.equator.com/en/latest/Getting_Started/) for setup instructions and basic Equator usage.

## Docs

Most documentation for Equator is located on [ReadTheDocs](http://docs.equator.com/en/latest/).

Some older documentation still exists on the [Wiki](https://github.com/equator/equator/wiki).

## Open-Source License

The Source Code Form of Equator is subject to the terms of the Mozilla Public
License, v. 2.0: [http://mozilla.org/MPL/2.0/](http://mozilla.org/MPL/2.0/)

The quick-and-dirty is you can do whatever if modifications to Equator are in
public GitHub forks. (Other ways to publicize modifications are also fine, as
are private use modifications. See also: [MPL 2.0 FAQ](https://www.mozilla.org/en-US/MPL/2.0/FAQ/))


CHANGELOG.md
## v0.10.1: Fix `font-size: 0` typing problems and more

_2016-03-21_

Important fix: remove `font-size: 0` on textarea (#585), fixing typing
in Chrome Canary (#540) as well as the Enter key not triggering the
`enter` handler in Webkit and Blink (#566). `transform: scale(0)` is
used instead and expected to be much more robust.

(Note: if you're coming from v0.9.x, there've been major API changes,
see the [v0.9.x → v0.10.0 Migration Guide][].)

[v0.9.x → v0.10.0 Migration Guide]: https://github.com/equator/equator/wiki/v0.9.x-%E2%86%92-v0.10.0-Migration-Guide

**new features:**
- (#544, #552, #558, #581) new symbols `\nparallel`, `\measuredangle`,
  `\odot`, `\parallelogram` (nonstandard), `\nless`, `\ngtr`, `\square`
- (#544) new commands `\overleftarrow`, `\overrightarrow`


**bugfixes:**
- (#585) fix typing in Chrome Canary, Enter key in Webkit+Blink
- (#582) fix `\degree` symbol to round-trip (rather than exporting
  `^\circ` which doesn't parse as one symbol)
- (#578) fix `.text()` to output `\cdot` as `*`
- (#529, #571, #574) fix `.text()` of fractions, spaces, variables followed
  by exponents
- (#577) fix `\triangle` symbol to match LaTeX better
- (#568) hotfix #435 order-dependence breaking clean build on Linux
- (#560) fix florin spacing still too close
- (#546) fix parsing or pasting `×` (Unicode times symbol)
- (#519/#487) fix auto-horizontal-scroll/pan on API calls
- (#528) fix #429 can't move cursor out of `TextBlock`
- (#526) fix exponentiation to export `^` not `**`
- (#525) fix Tab while there's a selection

**build system fixes:**
- (#532) add console output to show URL of local test pages

## v0.10.0: Total API overhaul, new features galore

_2016-02-20_

Many major changes including a total overhaul of the API (no more
auto-Equator-ifying of `.equator-editable` etc, and no more jQuery
plugin, instead global `Equator()` returns API objects, like jQuery
itself): See the [v0.9.x → v0.10.0 Migration Guide]
(https://github.com/equator/equator/wiki/v0.9.x-%E2%86%92-v0.10.0-Migration-Guide).

(If you already use the new global `Equator()`-based API from the
 `dev` branch, migrating to v0.10.0 should be just [one small change]
 (https://github.com/equator/equator/wiki/%60dev%60-branch-(2014%E2%80%932015)-%E2%86%92-v0.10.0-Migration-Guide)
 for you.)

**API-only changes:**
- (#336, #349, #351, #353) config options architecture
- (#308) don't auto-Equator-ify on jQuery `ready`
- (#297) prefix all CSS classes with `mq-`
- (#238, #272, #288, #337, #362, #459, #463, #495) kill jQuery plugin; new
  global `Equator()` returns API objects

**typist-facing changes:**
- (#506) delete `\caret` and `\underscore`
- (#453) incremental backspace: backspacing into a compound command like
  fraction or exponent goes left into it rather than selecting it
- (#285) render pasted text in math mode if cursor in math mode
- (5cf838d) LiveFraction (typing `/`) stops at space when expanding left
- (#264) intentional blur (like clicking outside field) clears selection
- (#262, #281, #391, #449, #509) auto-expanding, mis-matchable parens/pipes
- (#259) blue focus ring only around whole field not individual blocks
- (#258) `\sum` now comes with lower and upper limit blocks
- (#246, #248, #274, #434, #473) merge adjacent `SupSub`s into one
  command
- (#187) delete `\vector`
- (#144) Shift-Left/Right unselects back into a thing after selecting
  out of it
- (#157) stop fractions created by typing `/` at `,`/`;`/`:`

**new features:**
- (#468) add WOFF and WOFF2 font formats
- (#376, #398) add `autoSubscriptNumerals` option
- (#338) config option `sumStartsWithNEquals`
- (#321) static math instances may have `.innerFields`
- (#279) `leftRightIntoCmdGoes: 'up'/'down'`
- (#278, #407, #442) `SupSub` options to improve usability
- (#276, #410) anything focusable can be used to `substituteTextarea`
- (#263) typing `<=` and `>=` results in `\le` and `\ge`
- (#265) "autocommands": LaTeX control sequences that automatically
  render when you type the letters, without typing backslash first
- (#261, #361, #387, #404) when the math is too wide to fit in the
  field, pan/scroll horizontally
- (#247, #301, #255, #509) auto-unitalicize `sin`, `log` etc operator names
- (#245, #253) config option whether to Spacebar behaves like Tab
- (#241, #325, #425, #462) new API methods as used by Desmos
- (#191) `\class{classname}{math}` _a la_
  [MathJax](http://docs.mathjax.org/en/v2.2-latest/tex.html#html)
- (#151) `\textcolor{color}{math}`

**new build system features:**
- (#377) `OMIT_FONT_FACE=true make` omits `@font-face {...}`
- (#319) `make basic` builds stripped-down Equator for basic math

**bugfixes:**
- (#452) fix blinking blue cursor and autocorrect on iOS
- (#448) fix `\ddots` to be downward-rightward not upward-rightward
- (#432) fix quadratic-time fragment construction
- (#379) fix `.text()` errors when currently typing backslash command
- (#364, #367, #363, #397, #402, #417, #472) fixes to spacing and
  positioning
- (#323, #365, #409) fix LaTeX for `/`, `{`, `}` `^`, `_`, and `~`
- (99da82a) fix LaTeX parsing of `'`
- (#294, #355) fix `Cmd-Left` turns selection into typed text in Firefox
- (#296, #392) fix `f`/florin situation
- (#299) don't use reserved word `yield`
- (#284) escape non-ASCII Unicode characters in the JS source code
- (#272) fix API methods `.write()` on empty LaTeX and `.cmd()` erroring
- (#255) fix auto-spacing of `SupSub` and `PlusMinus`
- (#266) fix keyboard select after mouse select
- (#268) <code>\ </code> not `\:` as LaTeX for space
- (68c8f2b) fix resize gripper appearing sometimes in Chrome
- (6803077) fix Shift-Enter, Ctrl-Enter inputting newlines
- (f17fb95) fix potential Ctrl-C "copy" race condition
- (765dd70, #322) don't unnecessarily `stopPropagation()` mouse events
- (c1fe1ef, 9aef35f) fix up/down in an `\editable{}` in a fraction

**docs:**
- (#485) add more metadata to package.json
- (#484) fix links in README
- (#393) correctly credit co-creator @jneen

src/publicapi.js
/*********************************************************
 * The publicly exposed Equator API.
 ********************************************************/

var API = {}, Options = P(), optionProcessors = {}, Progenote = P(), EMBEDS = {};

/**
 * Interface Versioning (#459, #495) to allow us to virtually guarantee
 * backcompat. v0.10.x introduces it, so for now, don't completely break the
 * API for people who don't know about it, just complain with console.warn().
 *
 * The methods are shimmed in outro.js so that MQ.MathField.prototype etc can
 * be accessed.
 */
function insistOnInterVer() {
  if (window.console) console.warn(
    'You are using the Equator API without specifying an interface version, ' +
    'which will fail in v1.0.0. Easiest fix is to do the following before ' +
    'doing anything else:\n' +
    '\n' +
    '    Equator = Equator.getInterface(1);\n' +
    '    // now Equator.MathField() works like it used to\n' +
    '\n' +
    'See also the "`dev` branch (2014–2015) → v0.10.0 Migration Guide" at\n' +
    '  https://github.com/equator/equator/wiki/%60dev%60-branch-(2014%E2%80%932015)-%E2%86%92-v0.10.0-Migration-Guide'
  );
}
// globally exported API object
function Equator(el) {
  insistOnInterVer();
  return MQ1(el);
};
Equator.prototype = Progenote.p;
Equator.VERSION = "{VERSION}";
Equator.interfaceVersion = function(v) {
  // shim for #459-era interface versioning (ended with #495)
  if (v !== 1) throw 'Only interface version 1 supported. You specified: ' + v;
  insistOnInterVer = function() {
    if (window.console) console.warn(
      'You called Equator.interfaceVersion(1); to specify the interface ' +
      'version, which will fail in v1.0.0. You can fix this easily by doing ' +
      'this before doing anything else:\n' +
      '\n' +
      '    Equator = Equator.getInterface(1);\n' +
      '    // now Equator.MathField() works like it used to\n' +
      '\n' +
      'See also the "`dev` branch (2014–2015) → v0.10.0 Migration Guide" at\n' +
      '  https://github.com/equator/equator/wiki/%60dev%60-branch-(2014%E2%80%932015)-%E2%86%92-v0.10.0-Migration-Guide'
    );
  };
  insistOnInterVer();
  return Equator;
};
Equator.getInterface = getInterface;

var MIN = getInterface.MIN = 1, MAX = getInterface.MAX = 2;
function getInterface(v) {
  if (!(MIN <= v && v <= MAX)) throw 'Only interface versions between ' +
    MIN + ' and ' + MAX + ' supported. You specified: ' + v;

  /**
   * Function that takes an HTML element and, if it's the root HTML element of a
   * static math or math or text field, returns an API object for it (else, null).
   *
   *   var mathfield = MQ.MathField(mathFieldSpan);
   *   assert(MQ(mathFieldSpan).id === mathfield.id);
   *   assert(MQ(mathFieldSpan).id === MQ(mathFieldSpan).id);
   *
   */
  function MQ(el) {
    if (!el || !el.nodeType) return null; // check that `el` is a HTML element, using the
      // same technique as jQuery: https://github.com/jquery/jquery/blob/679536ee4b7a92ae64a5f58d90e9cc38c001e807/src/core/init.js#L92
    var blockId = $(el).children('.mq-root-block').attr(mqBlockId);
    var ctrlr = blockId && Node.byId[blockId].controller;
    return ctrlr ? APIClasses[ctrlr.KIND_OF_MQ](ctrlr) : null;
  };
  var APIClasses = {};

  MQ.L = L;
  MQ.R = R;
  MQ.saneKeyboardEvents = saneKeyboardEvents;

  function config(currentOptions, newOptions) {
    if (newOptions && newOptions.handlers) {
      newOptions.handlers = { fns: newOptions.handlers, APIClasses: APIClasses };
    }
    for (var name in newOptions) if (newOptions.hasOwnProperty(name)) {
      var value = newOptions[name], processor = optionProcessors[name];
      currentOptions[name] = (processor ? processor(value) : value);
    }
  }
  MQ.config = function(opts) { config(Options.p, opts); return this; };
  MQ.registerEmbed = function(name, options) {
    if (!/^[a-z][a-z0-9]*$/i.test(name)) {
      throw 'Embed name must start with letter and be only letters and digits';
    }
    EMBEDS[name] = options;
  };

  var AbstractEquator = APIClasses.AbstractEquator = P(Progenote, function(_) {
    _.init = function(ctrlr) {
      this.__controller = ctrlr;
      this.__options = ctrlr.options;
      this.id = ctrlr.id;
      this.data = ctrlr.data;
    };
    _.__mathquillify = function(classNames) {
      var ctrlr = this.__controller, root = ctrlr.root, el = ctrlr.container;
      ctrlr.createTextarea();

      var contents = el.addClass(classNames).contents().detach();
      root.jQ =
        $('<span class="mq-root-block"/>').attr(mqBlockId, root.id).appendTo(el);
      this.latex(contents.text());

      this.revert = function() {
        return el.empty().unbind('.equator')
        .removeClass('mq-editable-field mq-math-mode mq-text-mode')
        .append(contents);
      };
    };
    _.config = function(opts) { config(this.__options, opts); return this; };
    _.el = function() { return this.__controller.container[0]; };
    _.text = function() { return this.__controller.exportText(); };
    _.latex = function(latex) {
      if (arguments.length > 0) {
        this.__controller.renderLatexMath(latex);
        if (this.__controller.blurred) this.__controller.cursor.hide().parent.blur();
        return this;
      }
      return this.__controller.exportLatex();
    };
    _.html = function() {
      return this.__controller.root.jQ.html()
        .replace(/ equator-(?:command|block)-id="?\d+"?/g, '')
        .replace(/<span class="?mq-cursor( mq-blink)?"?>.?<\/span>/i, '')
        .replace(/ mq-hasCursor|mq-hasCursor ?/, '')
        .replace(/ class=(""|(?= |>))/g, '');
    };
    _.reflow = function() {
      this.__controller.root.postOrder('reflow');
      return this;
    };
  });
  MQ.prototype = AbstractEquator.prototype;

  APIClasses.EditableField = P(AbstractEquator, function(_, super_) {
    _.__mathquillify = function() {
      super_.__mathquillify.apply(this, arguments);
      this.__controller.editable = true;
      this.__controller.delegateMouseEvents();
      this.__controller.editablesTextareaEvents();
      return this;
    };
    _.focus = function() { this.__controller.textarea.focus(); return this; };
    _.blur = function() { this.__controller.textarea.blur(); return this; };
    _.write = function(latex) {
      this.__controller.writeLatex(latex);
      this.__controller.scrollHoriz();
      if (this.__controller.blurred) this.__controller.cursor.hide().parent.blur();
      return this;
    };
    _.empty = function() {
      var root = this.__controller.root, cursor = this.__controller.cursor;
      root.eachChild('postOrder', 'dispose');
      root.ends[L] = root.ends[R] = 0;
      root.jQ.empty();
      delete cursor.selection;
      cursor.insAtRightEnd(root);
      return this;
    };
    _.cmd = function(cmd) {
      var ctrlr = this.__controller.notify(), cursor = ctrlr.cursor;
      if (/^\\[a-z]+$/i.test(cmd) && !cursor.isTooDeep()) {
        cmd = cmd.slice(1);
        var klass = LatexCmds[cmd];
        if (klass) {
          cmd = klass(cmd);
          if (cursor.selection) cmd.replaces(cursor.replaceSelection());
          cmd.createLeftOf(cursor.show());
          this.__controller.scrollHoriz();
        }
        else /* TODO: API needs better error reporting */;
      }
      else cursor.parent.write(cursor, cmd);
      if (ctrlr.blurred) cursor.hide().parent.blur();
      return this;
    };
    _.select = function() {
      var ctrlr = this.__controller;
      ctrlr.notify('move').cursor.insAtRightEnd(ctrlr.root);
      while (ctrlr.cursor[L]) ctrlr.selectLeft();
      return this;
    };
    _.clearSelection = function() {
      this.__controller.cursor.clearSelection();
      return this;
    };

    _.moveToDirEnd = function(dir) {
      this.__controller.notify('move').cursor.insAtDirEnd(dir, this.__controller.root);
      return this;
    };
    _.moveToLeftEnd = function() { return this.moveToDirEnd(L); };
    _.moveToRightEnd = function() { return this.moveToDirEnd(R); };

    _.keystroke = function(keys) {
      var keys = keys.replace(/^\s+|\s+$/g, '').split(/\s+/);
      for (var i = 0; i < keys.length; i += 1) {
        this.__controller.keystroke(keys[i], { preventDefault: noop });
      }
      return this;
    };
    _.typedText = function(text) {
      for (var i = 0; i < text.length; i += 1) this.__controller.typedText(text.charAt(i));
      return this;
    };
    _.dropEmbedded = function(pageX, pageY, options) {
      var clientX = pageX - $(window).scrollLeft();
      var clientY = pageY - $(window).scrollTop();

      var el = document.elementFromPoint(clientX, clientY);
      this.__controller.seek($(el), pageX, pageY);
      var cmd = Embed().setOptions(options);
      cmd.createLeftOf(this.__controller.cursor);
    };
    _.clickAt = function(clientX, clientY, target) {
      target = target || document.elementFromPoint(clientX, clientY);

      var ctrlr = this.__controller, root = ctrlr.root;
      if (!jQuery.contains(root.jQ[0], target)) target = root.jQ[0];
      ctrlr.seek($(target), clientX + pageXOffset, clientY + pageYOffset);
      if (ctrlr.blurred) this.focus();
      return this;
    };
    _.ignoreNextMousedown = function(fn) {
      this.__controller.cursor.options.ignoreNextMousedown = fn;
      return this;
    };
  });
  MQ.EditableField = function() { throw "wtf don't call me, I'm 'abstract'"; };
  MQ.EditableField.prototype = APIClasses.EditableField.prototype;

  /**
   * Export the API functions that Equator-ify an HTML element into API objects
   * of each class. If the element had already been Equator-ified but into a
   * different kind (or it's not an HTML element), return null.
   */
  for (var kind in API) (function(kind, defAPIClass) {
    var APIClass = APIClasses[kind] = defAPIClass(APIClasses);
    MQ[kind] = function(el, opts) {
      var mq = MQ(el);
      if (mq instanceof APIClass || !el || !el.nodeType) return mq;
      var ctrlr = Controller(APIClass.RootBlock(), $(el), Options());
      ctrlr.KIND_OF_MQ = kind;
      return APIClass(ctrlr).__mathquillify(opts, v);
    };
    MQ[kind].prototype = APIClass.prototype;
  }(kind, API[kind]));

  return MQ;
}

Equator.noConflict = function() {
  window.Equator = origEquator;
  return Equator;
};
var origEquator = window.Equator;
window.Equator = Equator;

function RootBlockMixin(_) {
  var names = 'moveOutOf deleteOutOf selectOutOf upOutOf downOutOf'.split(' ');
  for (var i = 0; i < names.length; i += 1) (function(name) {
    _[name] = function(dir) { this.controller.handle(name, dir); };
  }(names[i]));
  _.reflow = function() {
    this.controller.handle('reflow');
    this.controller.handle('edited');
    this.controller.handle('edit');
  };
}
